torch/_inductor/runtime/benchmarking.py

```diff
@@ -0,0 +1,1001 @@
+            return other / self.timing_ms
+        return LazyBenchmark(lambda: other / self.timing_ms)
+
+
+def is_cpu_device(inputs: List[Any]) -> bool:
+    return all(
+        _input.device == torch.device("cpu")
+        for _input in inputs
+        if isinstance(_input, torch.Tensor)
+    )
+
+
+class Benchmarker:
+    def __init__(self: Self) -> None:
+        self.memory_cache: Dict[str, float] = {}
+        self.kwargs_hash_to_futures_gpu: Dict[
+            str, List[Tuple[Callable[[], Any], str]]
+        ] = {}
+
+    @cached_property
+    def L2_cache_size(self: Self) -> int:
+        counters["inductor"]["benchmarking_L2_cache_size"] += 1
+        device = torch.cuda.current_device()
+        properties = torch.cuda.get_device_properties(device)
+        return properties.L2_cache_size
+
+    @cached_property
+    def gpu_queue_limit(self: Self) -> int:
+        counters["inductor"]["benchmarking_gpu_queue_limit"] += 1
+        # ensures the queue is empty
+        torch.cuda.synchronize()
+        # capping the search space for the queue limit to 2500 is good enough,
+        # current queue limit on my machine (A100) is stable at ~1000, and
+        # in the event that we do artificially cap the queue limit to 2500
+        # we really shouldn't see any significant slowdowns
+        torch.cuda._sleep(
+            int(
+                (self.cpu_launch_overhead_ms_per_event_record * 2500)
+                / self.gpu_time_ms_per_gpu_clock_cycle
+            )
+        )
+        for idx in range(2500):
+            start_time_s = perf_counter()
+            torch.cuda.Event(enable_timing=True).record()
+            elapsed_time_ms = (perf_counter() - start_time_s) * 1000
+            # recording an event is near instantaneous, unless we have hit
+            # the queue limit, so 1ms seems like a good enough upper bound
+            if elapsed_time_ms > 1:
+                break
+        torch.cuda.synchronize()
+        return idx + 1
+
+    @cached_property
+    def cpu_launch_overhead_ms_per_event_record(self: Self) -> float:
+        counters["inductor"][
+            "benchmarking_cpu_launch_overhead_ms_per_event_record"
+        ] += 1
+        # ensures the queue is empty
+        torch.cuda.synchronize()
+        start_time_s = perf_counter()
+        for _ in range(100):
+            torch.cuda.Event(enable_timing=True).record()
+        torch.cuda.synchronize()
+        return ((perf_counter() - start_time_s) * 1000) / 100
+
+    @cached_property
+    def cpu_launch_overhead_ms_per_gpu_cache_clear(self: Self) -> float:
+        return self.cpu_launch_overhead_ms_and_gpu_time_ms_per_gpu_cache_clear[0]
+
+    @cached_property
+    def gpu_time_ms_per_gpu_cache_clear(self: Self) -> float:
+        return self.cpu_launch_overhead_ms_and_gpu_time_ms_per_gpu_cache_clear[1]
+
+    @cached_property
+    def cpu_launch_overhead_ms_and_gpu_time_ms_per_gpu_cache_clear(
+        self: Self,
+    ) -> Tuple[float, float]:
```

Comment:
nit, add docstrings on methods for more complicated ones

Reply:
done